Added attribute readers for rows and columns size

dsthode opened the pull request:
When iterating through the rows in a sheet, it would be nice to know the position relative to the row count, e.g. when you need to ignore the last row in a sheet. Columns size added as a convenience, too.

dsthode committed:
Added attribute readers for rows and columns size

dsthode committed:
Changed tabs to spaces

tomiacannondale commented:
Thank you for PR.

I check your code.
Please wait a little longer!

tomiacannondale commented:
Thank you for the PR.

Please add below test code to sheet_spec.rb and run rspec.

```
  describe "#row_size and #colum_size" do
    context "read general_datas" do
      include_context "read Sheet1"
      it { @sheet.row_size.should eq 16 }
      it { @sheet.column_size.should eq 3 }
    end

    context "read file created from excel 2010" do
      include_context "read file created from excel 2010"
      it { @sheet.row_size.should eq 4  }
      it { @sheet.column_size.should eq 1 }
    end
  end
```

I run rspec. 
`context "read general_datas"` are pass,
but `context "read file created from excel 2010"` are failed.

Expectation values are that seem to be contained data in `spec/data/created_from_excel2010.ods` or `spec/data/general_datas.ods`.

Let's think of another solution.

dsthode commented:
Thank you for testing the functions. 

I think the test results are right, because looking at the xml content of the Office 2010 sheet, it defines a number-rows-repeated value of 1048571 at the 5th row, and it defines another row after that, so it sums up to 1048577, just like the test result.

And for the columns, the sheet defines a number-columns-repeated of 1023 at the second column (at the last two rows, the number-columns-repeated has a value of 1024 at the first column), so that sums 1025.

Maybe that xml file should be manually edited to remove all that invisible clutter, or modify the test expectations to account for these invisible rows and columns.

tomiacannondale commented:
Thank you for comment.

I hope to be manually edited to remove all that invisible clutter.

And I came up with the following pattern.

| A | B | C | D | E |
| :-: | :-: | :-: | :-: | :-: |
| foo |  |  |  |  |
|  |  |  |  | YAY |
|  |  |  |  |  |
|  |  |  |  |  |
|  |  |  |  |  |
|  |  |  |  |  |
|  |  |  |  |  |
| baa |  |  |  |  |

How do you think about row size and colum size of the spread sheet?
I expect that `row_size` is 8, `column_size` is 5.

`number-rows-repeated` represents the same number of rows completely consecutive.
Therefore if spread sheet has blank lines are present in the middle,
Spread sheet input `number-rows-repeated` and value is repeated size.
Please see [Open Document Format for Office Applications (OpenDocument) Version 1.2#number-rows-repeated](http://docs.oasis-open.org/office/v1.2/os/OpenDocument-v1.2-os-part1.html#__RefHeading__1418532_253892949) 

Therefore `Sheet#each_row` is very complex.
I must refactor.

Solutions that I'm thinking of is to call `Sheet#each_row` and call `@cells.size` in `Sheet#row_size`.

dsthode commented:
I agree with the row and column size. I guess this table would me represented in a ODS file with a number-rows-repeated of 5 at the third row.
For each row, maybe the row numbers where number-rows-repeated is specified should be saved, so when there is an iteration on the rows, all the rows that fall in a number-rows-repeated row should return the same content. Something like this:

```
rows = []

rows[i] = {
   index: Number, 
   repeat-row: Number (from number-rows-repeated if present, otherwise 0), 
   content: RowContentFromODSFile
}
```

So, your previous example could be represented the following way:

```
rows[0] = {
   index: 0,
   repeat-row: 0,
   content: [foo, nil, nil, nil, nil]
}
rows[1] = {
   index: 1,
   repeat-row: 0,
   content: [nil, nil, nil, nil, YAY]
}
rows[2] = {
   index: 2,
   repeat-row: 5,
   content: [nil, nil, nil, nil, nil]
}
rows[3] = {
   index: 7,
   repeat-row: 0,
   content: [baa, nil, nil, nil, nil]
}
```

So when iterating on the rows, the number of iterations would be 8 (the row_size), and from iteration 3 to 7, row[2].content would be returned for each.
What do you think about this idea? I find it easy for rows, but for columns (with number-columns-repeated) looks a bit tougher.

tomiacannondale commented:
I feel your idea is very nice!!
Your idea is higher expandability and simple.
I think `number-columns-repeated` is same as the `number-rows-repteated`. Replace `content` inside `rows` to `columns` and `columns` has elemets that have same as `rows`.

```
rows[0] = {
   index: 0,
   repeat-row: 0,
   columns: [
     {index: 0,
      repeated-colum: 0,
      content: foo},
     {index: 1,
      repeated-colum: 3,
      content: nil}]
      },
rows[1] = {
   index: 1,
   repeat-row: 0,
   columns: [
     {index: 0,
      repeated-colum: 3,
      content: nil},
     {index: 1,
      repeated-colum: 0,
      content: YAY}]
},
rows[2] = {
   index: 2,
   repeat-row: 5,
   columns: [
     {index: 0,
      repeated-colum: 4,
      content: nil}]
},
rows[3] = {
   index: 7,
   repeat-row: 0,
   columns: [
     {index: 0,
      repeated-colum: 0,
      content: baa},
     {index: 1,
      repeated-colum: 3,
      content: nil}]
}
```

We can create another class above rows and columns.

Your guess is almost correct, but some parts are not correct.
Thired row has only `table:table-cell` that has number-columns-repeated which value is 4. Fourth row has number-rows-repeated, and value is 4.
Do you know how to convert ODS file to xml file? I converted to the following.

```
unzip <odsfile> -d <output-directory>
cd <output-directory>

nokogiri content.xml

# inside nokogiri command
> File.open(<xml-outputfile.xls>, 'w') {|f| @doc.puts f}
> exit
```

Then formated xml file is created.

dsthode commented:
Yes, you are right, the columns can be modified just like the rows, it's just a matter of creating the accessor method to keep in mind the columns-repeated.

Whenever I have needed to extract the xml contents of an ODS file I have followed the same procedure, unzip and copy the content.xml file.

tomiacannondale commented:
Yes, you are right. It is difficult to create the accessor method to keep mind the columns-repeated. rows-repeated too.

I came up with another way.

I will send PR to your repository. Please check my code.

dsthode committed:
Changed several thigs to enhance speed

dsthode committed:
Changed the way the sheet is initialized to improve performance when
iterating over the rows.
Also, added a new Row class to handle the virtual row numbering of the
Open Document Format when there is a numer-rows-repeated attribute on
any row.

dsthode commented:
Hi tomi, sorry for leaving this freeze for so long, but the project in which I was using your gem was delayed to get into other one more urgent. Now that the fire is gone, I have been able to retake the project where I left it.

I've made some changes to the gem to handle the rows the way I described above, that is: a new Row class that contains all the cells and the index of the row number_from and row number_to that corresponds to the row (because of the number-rows-repeated attribute) in the ODS file.

Also, I have changed the way the file is read and processed on the instantiation of the Sheet class, and now the contents are read and processed in the initialize method. I had problems with the way things were done before, the performance of the application was very bad with a file that had only 197 real rows, but because of the Open Document formatting, had 65537 "fake" rows and 255 "fake" columns. Iterating on these rows with each_rows brought my server to its knees because of all the XML processing, so I came with a different way, instantiating the rows on initialize.

I've changed too the accessor to the columns size, as I'm not really interested in this value, only on the number of rows. If you think it is useful, I can put it back.

Please, review the code and tell me what you think of it. It is being currently used here in my office.

tomiacannondale commented:
Sorry for late! I couldn't notice your change.

I will check your code as soon as.

and Please add test.

When will you use this code?
If you are not ready by using code, please tell me.

dsthode commented:
Don't worry I just pushed this into this PR today. I've run the current tests and they all terminate without issue, but maybe I could have a look and see if some more tests would be needed.
This code is currently being used in a pre-production application without issue, and as far as it goes, I have not found an error yet.

tomiacannondale commented on lib/rusk/sheet.rb:
```diff
@@ -21,89 +28,20 @@ def name= name
-        break if row_index > row
-        row_index += 1
+    def [](row_number)
+      rows.each do |r|
```
How about using Array#detect method?

``` ruby
rows.detect { |r| r.from >= row_number && row_number <= r.to }
```

tomiacannondale commented:
Thank you for commit.

This PR made Rusk::Sheet easy! Nice commit!

But, I pull your branch and run test. Tests failed. And I think that Rusk::Row(new class) should be added tests.
Please modify test and add tests of Rusk::Row.

Interface of Rusk::Sheet#[] was changed. ( `Rusk::Sheet[row, column]` to `Rusk::Sheet[row][column]` ) I think this change is not bad.

Test results are below.

```
bundle exec rspec
..................................FFFFFFFFF.FFFFFFFFFFFFFFFFFFF.FFFFFF.FFFFFFFFFFFFFFFFFFFFFF

Failures:

  1) Rusk::Cell modify ods file#value= 'mruby' changed to 'cruby'(string cell changed to string) 
     Failure/Error: @cell = set(0, 1, "cruby")
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/cell_spec.rb:218:in `block (5 levels) in <top (required)>'

  2) Rusk::Cell modify ods file#value= string cell changed to date 
     Failure/Error: @cell = set(0, 1, Date.new(2012,5,7))
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/cell_spec.rb:226:in `block (5 levels) in <top (required)>'

  3) Rusk::Cell modify ods file#value= string cell changed to date 
     Failure/Error: @cell = set(0, 1, Date.new(2012,5,7))
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/cell_spec.rb:226:in `block (5 levels) in <top (required)>'

  4) Rusk::Cell modify ods file#value= string cell changed to float 
     Failure/Error: @cell = set(0, 1, 500.1)
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/cell_spec.rb:235:in `block (5 levels) in <top (required)>'

  5) Rusk::Cell modify ods file#value= string cell changed to float 
     Failure/Error: @cell = set(0, 1, 500.1)
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/cell_spec.rb:235:in `block (5 levels) in <top (required)>'

  6) Rusk::Cell modify ods file#value= set to blank cell set string 
     Failure/Error: @cell = set(10, 0, "set string")
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/cell_spec.rb:245:in `block (6 levels) in <top (required)>'

  7) Rusk::Cell modify ods file#value= set to blank cell set string 
     Failure/Error: @cell = set(10, 0, "set string")
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/cell_spec.rb:245:in `block (6 levels) in <top (required)>'

  8) Rusk::Cell modify ods file#value= set to blank cell set float 
     Failure/Error: @cell = set(10, 0, 10)
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/cell_spec.rb:254:in `block (6 levels) in <top (required)>'

  9) Rusk::Cell modify ods file#value= set to blank cell set float 
     Failure/Error: @cell = set(10, 0, 10)
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/cell_spec.rb:254:in `block (6 levels) in <top (required)>'

  10) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
     Failure/Error: it { @sheet[2, 1].should be_kind_of Rusk::Cell }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:70:in `block (5 levels) in <top (required)>'

  11) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
     Failure/Error: it { @sheet[0, 0].value.should eq "string" }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:71:in `block (5 levels) in <top (required)>'

  12) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
     Failure/Error: it { @sheet[0, 1].value.should eq "mruby" }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:72:in `block (5 levels) in <top (required)>'

  13) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
     Failure/Error: it { @sheet[1, 1].value.should eq Date.new(2012,4,29) }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:73:in `block (5 levels) in <top (required)>'

  14) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
     Failure/Error: it { @sheet[4, 2].value.should eq 'after merged column' }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:74:in `block (5 levels) in <top (required)>'

  15) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
     Failure/Error: it { @sheet[5, 1].value.should eq 'merged first row cell' }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:75:in `block (5 levels) in <top (required)>'

  16) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
     Failure/Error: it { @sheet[6, 1].value.should eq 'merged second row cell' }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:76:in `block (5 levels) in <top (required)>'

  17) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
     Failure/Error: it { @sheet[8, 0].value.should eq "" }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:77:in `block (5 levels) in <top (required)>'

  18) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
     Failure/Error: it { @sheet[8, 1].value.should eq "" }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:78:in `block (5 levels) in <top (required)>'

  19) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
     Failure/Error: it { @sheet[11, 0].value.should eq 'hide the cell next to' }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:79:in `block (5 levels) in <top (required)>'

  20) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
     Failure/Error: it { @sheet[11, 1].value.should eq 'hidden cell' }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:80:in `block (5 levels) in <top (required)>'

  21) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with not exist cell index when row is existing, column is not existing 
     Failure/Error: it { @sheet[0, 1000].should be_nil }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:85:in `block (6 levels) in <top (required)>'

  22) Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with not exist cell index when row and column is not existing 
     Failure/Error: it { @sheet[1000, 10000].should be_nil }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:89:in `block (6 levels) in <top (required)>'

  23) Rusk::Sheet#[] read rows-repeated 
     Failure/Error: it { @sheet.map(&:value).should have(24).items }
     NoMethodError:
       undefined method `value' for #<Rusk::Row:0x007ffa3d3d7090>
     # ./lib/rusk/sheet.rb:43:in `block in each'
     # ./lib/rusk/sheet.rb:42:in `times'
     # ./lib/rusk/sheet.rb:42:in `each'
     # ./spec/sheet_spec.rb:96:in `map'
     # ./spec/sheet_spec.rb:96:in `block (4 levels) in <top (required)>'

  24) Rusk::Sheet#[] read rows-repeated 
     Failure/Error: it { @sheet.map(&:value).should eq @cells.flatten }
     NoMethodError:
       undefined method `value' for #<Rusk::Row:0x007ffa3d30f9f0>
     # ./lib/rusk/sheet.rb:43:in `block in each'
     # ./lib/rusk/sheet.rb:42:in `times'
     # ./lib/rusk/sheet.rb:42:in `each'
     # ./spec/sheet_spec.rb:97:in `map'
     # ./spec/sheet_spec.rb:97:in `block (4 levels) in <top (required)>'

  25) Rusk::Sheet#[] read created_from_excel2010 
     Failure/Error: it { @sheet[0, 0].value.should eq "created" }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:102:in `block (4 levels) in <top (required)>'

  26) Rusk::Sheet#[] read created_from_excel2010 
     Failure/Error: it { @sheet[1, 0].value.should eq "by" }
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:103:in `block (4 levels) in <top (required)>'

  27) Rusk::Sheet#[] read created_from_excel2010 
     Failure/Error: it { @sheet.map(&:value).should eq @cells.flatten }
     NoMethodError:
       undefined method `value' for #<Rusk::Row:0x007ffa3d5ae260>
     # ./lib/rusk/sheet.rb:43:in `block in each'
     # ./lib/rusk/sheet.rb:42:in `times'
     # ./lib/rusk/sheet.rb:42:in `each'
     # ./spec/sheet_spec.rb:104:in `map'
     # ./spec/sheet_spec.rb:104:in `block (4 levels) in <top (required)>'

  28) Rusk::Sheet#each access order by row and column
     Failure/Error: cell.value.should eq cells[index]
     NoMethodError:
       undefined method `value' for #<Rusk::Row:0x007ffa3d554e68>
     # ./spec/sheet_spec.rb:115:in `block (4 levels) in <top (required)>'
     # ./lib/rusk/sheet.rb:43:in `block in each'
     # ./lib/rusk/sheet.rb:42:in `times'
     # ./lib/rusk/sheet.rb:42:in `each'
     # ./spec/sheet_spec.rb:114:in `block (3 levels) in <top (required)>'

  29) Rusk::Sheet#each_row read Sheet1 of general_datas.ods behaves like Sheet#each_row access order by first row access order by first row
     Failure/Error: @sheet.each_row do |rows|
     NoMethodError:
       undefined method `each_row' for #<Rusk::Sheet:0x007ffa3d3d9138>
     Shared Example Group: "Sheet#each_row access order by first row" called from ./spec/sheet_spec.rb:142
     # ./spec/sheet_spec.rb:131:in `block (4 levels) in <top (required)>'

  30) Rusk::Sheet#each_row read file created from excel behaves like Sheet#each_row access order by first row access order by first row
     Failure/Error: @sheet.each_row do |rows|
     NoMethodError:
       undefined method `each_row' for #<Rusk::Sheet:0x007ffa3d50b178>
     Shared Example Group: "Sheet#each_row access order by first row" called from ./spec/sheet_spec.rb:147
     # ./spec/sheet_spec.rb:131:in `block (4 levels) in <top (required)>'

  31) Rusk::Sheet#each_row read file created from excel once call Sheet#each_row behaves like Sheet#each_row access order by first row access order by first row
     Failure/Error: @sheet.each_row do |cell|
     NoMethodError:
       undefined method `each_row' for #<Rusk::Sheet:0x007ffa3d530c70>
     Shared Example Group: "Sheet#each_row access order by first row" called from ./spec/sheet_spec.rb:155
     # ./spec/sheet_spec.rb:151:in `block (5 levels) in <top (required)>'

  32) Rusk::Sheet#each_row without block 
     Failure/Error: it { @sheet.each_row.should be_kind_of Enumerator }
     NoMethodError:
       undefined method `each_row' for #<Rusk::Sheet:0x007ffa3d5a00c0>
     # ./spec/sheet_spec.rb:161:in `block (4 levels) in <top (required)>'

  33) Rusk::Sheet#each_column access order by first column
     Failure/Error: @sheet.each_column do |columns|
     NoMethodError:
       undefined method `each_column' for #<Rusk::Sheet:0x007ffa3d596d18>
     # ./spec/sheet_spec.rb:170:in `block (3 levels) in <top (required)>'

  34) Rusk::Sheet#each_column without block 
     Failure/Error: it { @sheet.each_column.should be_kind_of Enumerator }
     NoMethodError:
       undefined method `each_column' for #<Rusk::Sheet:0x007ffa3d4f7510>
     # ./spec/sheet_spec.rb:178:in `block (4 levels) in <top (required)>'

  35) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  36) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  37) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  38) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  39) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  40) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  41) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  42) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  43) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  44) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  45) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  46) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  47) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  48) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  49) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  50) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  51) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  52) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  53) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  54) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  55) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

  56) Rusk::Sheet modify ods file read rows-repeated 
     Failure/Error: sheet[1, 1].value = "modified 1,1"
     ArgumentError:
       wrong number of arguments (2 for 1)
     # ./lib/rusk/sheet.rb:31:in `[]'
     # ./spec/sheet_spec.rb:209:in `block (5 levels) in <top (required)>'
     # ./lib/rusk/book.rb:15:in `initialize'
     # ./lib/rusk/book.rb:36:in `new'
     # ./lib/rusk/book.rb:36:in `open'
     # ./spec/sheet_spec.rb:207:in `block (4 levels) in <top (required)>'

Finished in 0.57807 seconds
93 examples, 56 failures

Failed examples:

rspec ./spec/cell_spec.rb:221 # Rusk::Cell modify ods file#value= 'mruby' changed to 'cruby'(string cell changed to string) 
rspec ./spec/cell_spec.rb:229 # Rusk::Cell modify ods file#value= string cell changed to date 
rspec ./spec/cell_spec.rb:230 # Rusk::Cell modify ods file#value= string cell changed to date 
rspec ./spec/cell_spec.rb:238 # Rusk::Cell modify ods file#value= string cell changed to float 
rspec ./spec/cell_spec.rb:239 # Rusk::Cell modify ods file#value= string cell changed to float 
rspec ./spec/cell_spec.rb:248 # Rusk::Cell modify ods file#value= set to blank cell set string 
rspec ./spec/cell_spec.rb:249 # Rusk::Cell modify ods file#value= set to blank cell set string 
rspec ./spec/cell_spec.rb:257 # Rusk::Cell modify ods file#value= set to blank cell set float 
rspec ./spec/cell_spec.rb:258 # Rusk::Cell modify ods file#value= set to blank cell set float 
rspec ./spec/sheet_spec.rb:70 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
rspec ./spec/sheet_spec.rb:71 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
rspec ./spec/sheet_spec.rb:72 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
rspec ./spec/sheet_spec.rb:73 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
rspec ./spec/sheet_spec.rb:74 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
rspec ./spec/sheet_spec.rb:75 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
rspec ./spec/sheet_spec.rb:76 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
rspec ./spec/sheet_spec.rb:77 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
rspec ./spec/sheet_spec.rb:78 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
rspec ./spec/sheet_spec.rb:79 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
rspec ./spec/sheet_spec.rb:80 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with exist cell index 
rspec ./spec/sheet_spec.rb:85 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with not exist cell index when row is existing, column is not existing 
rspec ./spec/sheet_spec.rb:89 # Rusk::Sheet#[] read Sheet1(merged, columns-repeated, various data type) with not exist cell index when row and column is not existing 
rspec ./spec/sheet_spec.rb:96 # Rusk::Sheet#[] read rows-repeated 
rspec ./spec/sheet_spec.rb:97 # Rusk::Sheet#[] read rows-repeated 
rspec ./spec/sheet_spec.rb:102 # Rusk::Sheet#[] read created_from_excel2010 
rspec ./spec/sheet_spec.rb:103 # Rusk::Sheet#[] read created_from_excel2010 
rspec ./spec/sheet_spec.rb:104 # Rusk::Sheet#[] read created_from_excel2010 
rspec ./spec/sheet_spec.rb:111 # Rusk::Sheet#each access order by row and column
rspec ./spec/sheet_spec.rb:129 # Rusk::Sheet#each_row read Sheet1 of general_datas.ods behaves like Sheet#each_row access order by first row access order by first row
rspec ./spec/sheet_spec.rb:129 # Rusk::Sheet#each_row read file created from excel behaves like Sheet#each_row access order by first row access order by first row
rspec ./spec/sheet_spec.rb:129 # Rusk::Sheet#each_row read file created from excel once call Sheet#each_row behaves like Sheet#each_row access order by first row access order by first row
rspec ./spec/sheet_spec.rb:161 # Rusk::Sheet#each_row without block 
rspec ./spec/sheet_spec.rb:167 # Rusk::Sheet#each_column access order by first column
rspec ./spec/sheet_spec.rb:178 # Rusk::Sheet#each_column without block 
rspec ./spec/sheet_spec.rb:220 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:221 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:222 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:223 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:224 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:225 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:226 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:227 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:228 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:229 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:230 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:231 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:232 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:233 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:234 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:235 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:236 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:237 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:238 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:239 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:240 # Rusk::Sheet modify ods file read rows-repeated 
rspec ./spec/sheet_spec.rb:241 # Rusk::Sheet modify ods file read rows-repeated
```

dsthode commented:
Thank you very much.

Respecting Array#detect, it's true, sometimes I miss the Ruby idioms, will change that in the next commit.

You are right about tests, they fail, I don't really know how I run them. I'll fix them and I'll add a test file for the Row class.

tomiacannondale commented:
OK.
Rusk use [rspec](https://github.com/rspec/rspec) for testing library.
I often read [rspec relish](https://relishapp.com/rspec/).
To develop rusk, I recommend to use [bundler](http://bundler.io).
Bundler install necessary libraries to develop rusk.

How to run test below.

```
bundle install --path vendor/bundle  # if did not use bundler(only once)
bundle exec rspec                    # run test
```

I am looking forward to your commit.

tomiacannondale commented:
I will close this PR. Because This PR is not not progressing discussions.
If you want to continue the PR, please add comment.
I will close on next Saturday(2015.2.14 +09:00).
